File: LC1448.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxCount(self, root, x):
        if not root:
            return
        
        if root.val >= x:
            x = root.val
            self.count += 1
        
        if root.left:
            self.maxCount(root.left, x)
        
        if root.right:
            self.maxCount(root.right, x)

    def goodNodes(self, root: TreeNode) -> int:
        self.count = 0

        self.maxCount(root, root.val)

        return self.count

File: LC1448.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {

private:
    int count = 0;
public:

    void maxCount(TreeNode* root, int x){
        if(root->val >= x){
            x = root->val;
            count++;
        }
        
        if(root->left)
            maxCount(root->left, x);
        
        if(root->right)
            maxCount(root->right, x);

    }

    int goodNodes(TreeNode* root) {
        if(!root)
            return 0;

        maxCount(root, root->val);

        return count;
    }
};
